<a href="https://colab.research.google.com/github/lauBit/ML-taller2/blob/main/Taller_1_Grupo_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Taller 1 - Grupo 6

In [ ]:
# Instalación de bibliotecas necesarias
pip install mlxtend

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Instalación de bibliotecas necesarias
!pip install scikit-surprise

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Importación de bibliotecas
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules, fpgrowth, fpmax
from mlxtend.preprocessing import TransactionEncoder
from surprise import Dataset, Reader, KNNBasic, KNNBaseline
from surprise.model_selection import train_test_split
from surprise import accuracy
from surprise.model_selection import KFold
from surprise import Dataset, Reader

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Carga del conjunto de datos
df= pd.read_csv('/content/steam-200k.csv')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Visualización de las primeras 10 filas del DataFrame
df.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


151603712  The Elder Scrolls V Skyrim  purchase    1.0  0
0  151603712  The Elder Scrolls V Skyrim      play  273.0  0
1  151603712                   Fallout 4  purchase    1.0  0
2  151603712                   Fallout 4      play   87.0  0
3  151603712                       Spore  purchase    1.0  0
4  151603712                       Spore      play   14.9  0
5  151603712           Fallout New Vegas  purchase    1.0  0
6  151603712           Fallout New Vegas      play   12.1  0
7  151603712               Left 4 Dead 2  purchase    1.0  0
8  151603712               Left 4 Dead 2      play    8.9  0
9  151603712                    HuniePop  purchase    1.0  0

In [ ]:
# Renombrar columnas del DataFrame
df.columns = ['user_id', 'game-title', 'behavior-name', 'value','valor_0']

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


user_id                  game-title behavior-name  value  valor_0
0       151603712  The Elder Scrolls V Skyrim          play  273.0        0
1       151603712                   Fallout 4      purchase    1.0        0
2       151603712                   Fallout 4          play   87.0        0
3       151603712                       Spore      purchase    1.0        0
4       151603712                       Spore          play   14.9        0
...           ...                         ...           ...    ...      ...
199994  128470551                 Titan Souls          play    1.5        0
199995  128470551  Grand Theft Auto Vice City      purchase    1.0        0
199996  128470551  Grand Theft Auto Vice City          play    1.5        0
199997  128470551                        RUSH      purchase    1.0        0
199998  128470551                        RUSH          play    1.4        0

[199999 rows x 5 columns]

In [ ]:
# Crear una tabla dinámica para reorganizar los datos
df_pivoted = df.pivot_table(index=['user_id', 'game-title'],
                            columns='behavior-name',
                            values='value',
                            aggfunc='first').reset_index()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Eliminar el nombre de las columnas
df_pivoted.columns.name = None

# Renombrar columnas específicas
df_pivoted.rename(columns={'purchase': 'purchase_value', 'play': 'play_time'}, inplace=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Visualización de las primeras 10 filas del DataFrame pivotado
df_pivoted.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


user_id                game-title  play_time  purchase_value
0     5250               Alien Swarm        4.9             1.0
1     5250           Cities Skylines      144.0             1.0
2     5250            Counter-Strike        NaN             1.0
3     5250     Counter-Strike Source        NaN             1.0
4     5250             Day of Defeat        NaN             1.0
5     5250        Deathmatch Classic        NaN             1.0
6     5250  Deus Ex Human Revolution       62.0             1.0
7     5250                    Dota 2        0.2             1.0
8     5250                 Half-Life        NaN             1.0
9     5250               Half-Life 2        NaN             1.0

In [ ]:
# Descripción del DataFrame pivotado
df_pivoted.describe()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


user_id     play_time  purchase_value
count  1.288040e+05  70477.000000        128803.0
mean   1.025343e+08     48.886105             1.0
std    7.242875e+07    229.353927             0.0
min    5.250000e+03      0.100000             1.0
25%    4.559264e+07      1.000000             1.0
50%    8.605570e+07      4.500000             1.0
75%    1.542307e+08     19.100000             1.0
max    3.099031e+08  11754.000000             1.0

In [ ]:
# Información del DataFrame pivotado
df_pivoted.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128804 entries, 0 to 128803
Data columns (total 4 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   user_id         128804 non-null  int64  
 1   game-title      128804 non-null  object 
 2   play_time       70477 non-null   float64
 3   purchase_value  128803 non-null  float64
dtypes: float64(2), int64(1), object(1)
memory usage: 3.9+ MB


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Filtrar y depurar el DataFrame
df_pivoted_depurado = df_pivoted[['user_id', 'game-title', 'play_time']].dropna()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Ajustar los valores de tiempo de juego ajustando los datos logaritmicamente
df_pivoted_depurado['play_time_ajustado'] = df_pivoted_depurado['play_time'].apply(lambda x: np.log1p(x) if x > 0 else 0)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
df_pivoted_depurado.describe()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


user_id     play_time  play_time_ajustado
count  7.047700e+04  70477.000000        70477.000000
mean   1.058852e+08     48.886105            2.053191
std    7.150551e+07    229.353927            1.615661
min    5.250000e+03      0.100000            0.095310
25%    4.934256e+07      1.000000            0.693147
50%    9.169067e+07      4.500000            1.704748
75%    1.556738e+08     19.100000            3.000720
max    3.099031e+08  11754.000000            9.372034

In [ ]:
# Crear un DataFrame limpio con las columnas necesarias
df_limpio = df_pivoted_depurado[['user_id', 'game-title', 'play_time_ajustado']]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Definir el rango de calificación
reader = Reader(rating_scale=(df_limpio['play_time_ajustado'].min(), df_limpio['play_time_ajustado'].max()))

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Cargar los datos en el formato adecuado
data = Dataset.load_from_df(df_limpio, reader)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Dividir los datos en conjuntos de entrenamiento y prueba
trainset, testset = train_test_split(data, test_size=0.25, random_state=123)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
print(trainset.n_items)

3297


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Inicializar el modelo KNN
knn = KNNBasic()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Ajustar el modelo KNN al conjunto de entrenamiento
knn.fit(trainset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Computing the msd similarity matrix...
Done computing similarity matrix.


In [ ]:
# Opciones de similitud del modelo KNN
knn.sim_options

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{'user_based': True}

In [ ]:
# Predecir sobre el conjunto de prueba
predictions = knn.test(testset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Visualizar las predicciones
predictions

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[Prediction(uid=30061970, iid='Banished', r_ui=1.0986122886681096, est=2.351477286779714, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=107097425, iid='Team Fortress 2', r_ui=3.4965075614664802, est=3.2311234786708636, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=211920439, iid='Dota 2', r_ui=3.4011973816621555, est=2.05130638336984, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid=159428605, iid='The Lord of the Rings War in the North', r_ui=2.6878474937846906, est=2.0680725287272557, details={'actual_k': 22, 'was_impossible': False}),
 Prediction(uid=51557405, iid='GRID 2', r_ui=2.8154087194227095, est=2.164234704485483, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=174907851, iid='FINAL FANTASY XIV A Realm Reborn', r_ui=3.784189633918261, est=3.349887179563313, details={'actual_k': 3, 'was_impossible': False}),
 Prediction(uid=176539817, iid='Unturned', r_ui=3.8918202981106

In [ ]:
# Calcular y mostrar el RMSE de las predicciones
accuracy.rmse(predictions)

RMSE: 1.4337


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


1.4337099599171463

## **Algoritmo SVD**

In [ ]:
# Algoritmo SVD para comparación
from surprise import SVD
from surprise.model_selection import cross_validate

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Inicializar el modelo SVD
svd = SVD()

# Ajustar el modelo SVD al conjunto de entrenamiento
svd.fit(trainset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Validación cruzada del modelo SVD
results = cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Imprimir resultados de RMSE y MAE
print('RMSE:', results['test_rmse'].mean())
print('MAE:', results['test_mae'].mean())

RMSE: 1.3997353362965197
MAE: 1.088633776002097


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Predecir sobre el conjunto de prueba con el modelo SVD
predictions_1 = svd.test(testset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Calcular y mostrar el RMSE de las predicciones SVD
accuracy.rmse(predictions_1)

RMSE: 0.9126


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0.9126070288577363

In [ ]:
df_limpio.describe()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


user_id  play_time_ajustado
count  7.047700e+04        70477.000000
mean   1.058852e+08            2.053191
std    7.150551e+07            1.615661
min    5.250000e+03            0.095310
25%    4.934256e+07            0.693147
50%    9.169067e+07            1.704748
75%    1.556738e+08            3.000720
max    3.099031e+08            9.372034

In [ ]:
predictions_1

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[Prediction(uid=30061970, iid='Banished', r_ui=1.0986122886681096, est=1.6579417773931056, details={'was_impossible': False}),
 Prediction(uid=107097425, iid='Team Fortress 2', r_ui=3.4965075614664802, est=2.6568161615484582, details={'was_impossible': False}),
 Prediction(uid=211920439, iid='Dota 2', r_ui=3.4011973816621555, est=3.269885819266494, details={'was_impossible': False}),
 Prediction(uid=159428605, iid='The Lord of the Rings War in the North', r_ui=2.6878474937846906, est=1.790011888458773, details={'was_impossible': False}),
 Prediction(uid=51557405, iid='GRID 2', r_ui=2.8154087194227095, est=2.129331419272071, details={'was_impossible': False}),
 Prediction(uid=174907851, iid='FINAL FANTASY XIV A Realm Reborn', r_ui=3.784189633918261, est=2.938468638384568, details={'was_impossible': False}),
 Prediction(uid=176539817, iid='Unturned', r_ui=3.8918202981106265, est=2.546316310524734, details={'was_impossible': False}),
 Prediction(uid=233726801, iid='Dota 2', r_ui=1.3609765

# Conclusiones

Se realizaron dos sistemas de recomendación: KNNBasic y SVD.

Se evaluó el rendimiento (RMSE) de cada uno de los modelos obteniendo los siguientes resultados:
- KNN: 1.4337
- SVD: 0.9126

De acuerdo con lo anterior, se concluye que el modelo SVD generó el mejor resultado.